### DISCUSSION TITLE
Confused with how @apply order works

### DISCUSSION THREAD

**Discussion by marco-carvalho** [tailwindlabs/tailwindcss#7873] [Help] [answered]
When I specify this custom class:
```
.foo1 {
  @apply border-r-2 border-r-4 border-r-8
}
```
Tailwind generates this CSS:
```
.foo1 {
    border-right-width: 8px
}
```
And Tailwind even advises me:
```
'border-r-2' applies the same CSS property as 'border-r-4' and 'border-r-8'
```
Which leads me to think that Tailwind ignores the previous redundant ones and uses the style from the last one.
___
But when I specify this custom class:
```
.foo2 {
  @apply border-r-8 border-x-2 
}
```
Tailwind generates this CSS:
```
.foo2 {
    border-left-width: 2px;
    border-right-width: 8px
}
```
But I thought it should generates this CSS:
```
.foo2 {
    border-left-width: 2px;
    border-right-width: 2px;
}
```
Because `border-r-8` applies the same CSS property `border-right-width` as `border-x-2`
___
Am I getting something wrong with how `@apply` works?

**Comment by adamwathan** [comment, marked as answer]
The main thing to understand about `@apply` is that it's designed to behave exactly the same as putting those classes in your HTML.

So given your example here:

```css
.foo2 {
  @apply border-r-8 border-x-2 
}
```

...if you write this HTML:

```html
<div class="border-r-8 border-x-2">...</div>
```

...you get an element with a 2px border on the left and an 8px border on the right:

https://play.tailwindcss.com/9Ix55ruUYg

So applying those classes needs to behave the same, otherwise extracting those classes with `@apply` isn't a true refactoring.

Classes are _not_ applied from left to right, they are applied based on the order those classes appear individually in the generated CSS.

So these are identical, because the order you type them in doesn't matter:

```css
.foo1 {
  @apply border-x-2 border-r-8 
}
.foo2 {
  @apply border-r-8 border-x-2 
}
```

Now in this case, that doesn't quite apply:

```css
.foo1 {
  @apply border-r-2 border-r-4 border-r-8
}

.foo2 {
  @apply border-r-8 border-r-4 border-r-2
}
```

This is because all of these classes are direct collisions (which is why we warn about them) and classes that are direct collisions have a non-deterministic sort order in the CSS output in Tailwind for performance reasons — it's faster to not sort them since they should never be used at the same time anyways so the declaration order within the set of collisions doesn't matter.

Hope that clears things up a little bit!

TL;DR: Don't `@apply` colliding classes, we consider that an error, and outside of collisions the order of classes has no impact on the generated CSS, just like the class attribute in HTML.

**Reply by marco-carvalho** [reply]
thanks @adamwathan
